Notebook on accelerating deep neural networks with tensorly

jacobgil opened the pull request:
A notebook on accelerating deep neural networks with TensorLy and PyTorch

JeanKossaifi commented on 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb:
```diff
@@ -0,0 +1,594 @@
+    "1. Faster.\n",
+    "2. Smaller.\n",
+    "\n",
+    "Like pruning, these are tools that can be used to make the networks smaller and faster, and then try to restore accuracy by training further.\n",
```
Typo: (decomositions instead of decomposition) 
Also the two sentences are slightly redundant, we could replace by a similar paragraph such as:

We will show how to use Tucker and CP decompositions to find low rank approximations of convolutional layers.
Akin to prunning, this will allow use to reduce the number of parameters as well as make them faster. While the low-rank approximation can decrease the accuracy, this can be restored by training further.

JeanKossaifi commented on 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb:
```diff
@@ -0,0 +1,594 @@
+    "\n",
+    "---\n",
+    "\n",
+    "First lets get VGG19 from torch vision: "
```
Let's load a pretrained VGG-19 network, using torch vision?
Also, it might be clearer to replace the `print(model)` by a short text explaining how the VGG net works (e.g. buch of conv, relu, pooling followed by two large FC layers before the final classification layer)

JeanKossaifi commented on 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb:
```diff
@@ -0,0 +1,594 @@
+    "Plugging this into the formula for the convolutional layer output form above:\n",
+    "\n",
+    "$ V(x, y, t) = \\sum_r\\sum_i \\sum_j \\sum_sK^x_r(i)K^y_r(i)K^s_r(s)K^t_r(t)X(x-i, y-j, s) $\n",
+    "$ = \\sum_rK^t_r(t) \\sum_i \\sum_j K^x_r(i)K^y_r(j)\\sum_sK^s_r(s)X(x-i, y-j, s) $ \n",
```
I guess later on we can add some images and equivalent formulation to make the whole formula more human friendly :)

JeanKossaifi commented on 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb:
```diff
@@ -0,0 +1,594 @@
+    "import operator\n",
+    "\n",
+    "def count_params(model):    \n",
+    "    return  sum(reduce(operator.mul, p.size(), 1) for p in model.parameters())\n",
```
since we use numpy already, we could use `np.sum` instead of `sum` and `np.prod` instead of the reduce on mul. This saves us from importing reduce and operator and might be slightly clearer to read :)

jacobgil commented:
Thanks,
I will address the comments, move it to the PyTorch backend and add an example of running on an image.
Maybe after that we can merge as a first version.

I was also thinking about adding code for demonstrating fine tuning (like in my original repo that does that on a cat/dog classifier) to restore accuracy, 
But that means a lot of more additional code will need to be imported to the notebook, and I thought maybe it would become less clear. 
I wanted to avoid a situation where the actual code that does the decomposition would become the minority :-)

jacobgil commented:
@JeanKossaifi 
- I addressed the comments.
- Switched to the PyTorch backend (and tested with Python 3).
- Forward pass on an image before and after the decomposition, to show that the network is still sensible.

I guess next big potential changes would be:
- Adding auto rank selection with VBMF.
- Fine tuning.
But i'm not sure if it will over complicate the notebook.

jacobgil commented:
@JeanKossaifi 
Let me know if you think something should be changed for a merge.

JeanKossaifi commented:
Awesome, thanks @jacobgil!

I agree, let's merge this and we can think on adding more / having a follow-up notebook! :)
